tests/test_derivatives.py

```diff
@@ -63,6 +69,40 @@ def test_stencil_derivative(self, SymbolType, dim):
         assert(np.allclose(u_di.data, 66.6))
         assert(np.allclose(u_dii.data, 66.6))
 
+    @pytest.mark.parametrize('SymbolType, derivative, dim, expected', [
+        (Function, ['dx2'], 3, 'Derivative(u(x, y, z), (x, 2))'),
+        (Function, ['dx2dy'], 3, 'Derivative(u(x, y, z), (x, 2), y)'),
+        (Function, ['dx2dydz'], 3, 'Derivative(u(x, y, z), (x, 2), y, z)'),
+        (Function, ['dx2', 'dy'], 3, 'Derivative(Derivative(u(x, y, z), (x, 2)), y)'),
+        (Function, ['dx2dy', 'dz2'], 3,
+         'Derivative(Derivative(u(x, y, z), (x, 2), y), (z, 2))'),
+        (TimeFunction, ['dx2'], 3, 'Derivative(u(t, x, y, z), (x, 2))'),
+        (TimeFunction, ['dx2dy'], 3, 'Derivative(u(t, x, y, z), (x, 2), y)'),
+        (TimeFunction, ['dx2', 'dy'], 3,
+         'Derivative(Derivative(u(t, x, y, z), (x, 2)), y)')
+    ])
+    def test_unevaluation(self, SymbolType, derivative, dim, expected):
+        u = SymbolType(name='u', grid=self.grid, time_order=2, space_order=2)
+        expr = getattr(u, derivative[0])
+        for d in derivative[1:]:
+            expr = getattr(expr, d)
+        assert(expr.__str__() == expected)
+        # Make sure the FD evaluation executes
+        expr.evaluate
+
+    @pytest.mark.parametrize('expr,expected', [
+        ('u.dx + u.dy', 'Derivative(u, x) + Derivative(u, y)'),
+        ('u.laplace',
+         'Derivative(u, (x, 2)) + Derivative(u, (y, 2)) + Derivative(u, (z, 2))'),
+        ('(u.dx*u.dy).dx', 'Derivative(Derivative(u, x) * Derivative(u, y), x)')
+    ])
+    def test_arithmetic(self, expr, expected):
```

Comment:
Can we add more test cases? 

For example

```
(u.dx + u.dy).dx
((u.dx + u.dy).dx + u.dxdy).dx
(u**4).dx
(u / 4).dx
((u.dx + v.dy).dx * v.dx).dy. # yeah, could do arithmetic with another Function too
```

etc. I think this **really needs to be bullet proof**, otherwise things can get screwed in the most unpredictable ways!

Reply:
Sure

Reply:
Done